bindings/pyroot_experimental/PyROOT/src/TClonesArrayPyz.cxx

```diff
@@ -0,0 +1,159 @@
+// Author: Enric Tejedor CERN  02/2019
+// Original PyROOT code by Wim Lavrijsen, LBL
+
+/*************************************************************************
+ * Copyright (C) 1995-2019, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+// Bindings
+#include "CPyCppyy.h"
+#include "PyROOTPythonize.h"
+#include "CPPInstance.h"
+#include "MemoryRegulator.h"
+#include "Utility.h"
+#include "PyzCppHelpers.hxx"
+
+// ROOT
+#include "TClass.h"
+#include "TClonesArray.h"
+#include "TBufferFile.h"
+
+using namespace CPyCppyy;
+
+// Clone an object into a position of a TClonesArray
+static TObject *CloneObjectInPlace(const TObject *obj, TClonesArray *cla, int index)
+{
+   // Create object with default constructor at index
+   char *arrObj = (char *)cla->New(index);
+   if (!arrObj) {
+      PyErr_Format(PyExc_RuntimeError, "Failed to create new object at index %d of TClonesArray", index);
+      return nullptr;
+   }
+
+   auto baseOffset = obj->IsA()->GetBaseClassOffset(TObject::Class());
+   auto newObj = (TObject *)(arrObj + baseOffset);
+
+   // Create a buffer where the object will be streamed
+   TBufferFile buffer(TBuffer::kWrite, cla->GetClass()->Size());
+   buffer.MapObject(obj); // register obj in map to handle self reference
+   const_cast<TObject *>(obj)->Streamer(buffer);
+
+   // Read new object from buffer
+   buffer.SetReadMode();
+   buffer.ResetMap();
+   buffer.SetBufferOffset(0);
+   buffer.MapObject(newObj); // register obj in map to handle self reference
+   newObj->Streamer(buffer);
+   newObj->ResetBit(kIsReferenced);
+   newObj->ResetBit(kCanDelete);
+
+   return newObj;
+}
+
+// Convert Python index into straight C index
+static PyObject *PyStyleIndex(PyObject *self, PyObject *index)
+{
+   Py_ssize_t idx = PyInt_AsSsize_t(index);
+   if (idx == (Py_ssize_t)-1 && PyErr_Occurred())
+      return nullptr;
+
+   // To know the capacity of a TClonesArray, we need to invoke GetSize
+   PyObject *pysize = CallPyObjMethod(self, "GetSize");
+   if (!pysize) {
+      PyErr_Clear();
+      return nullptr;
+   }
+
+   Py_ssize_t size = PyInt_AsSsize_t(pysize);
+   Py_DECREF(pysize);
+   if (idx >= size || (idx < 0 && idx < -size)) {
+      PyErr_SetString(PyExc_IndexError, "index out of range");
+      return nullptr;
+   }
+
+   PyObject *pyindex = nullptr;
```

Comment:
Do you need to increment the count here? You could just do a pass-through?

Reply:
The caller of `PyStyleIndex` is responsible for decrefing the return value (`pyindex`). `pyindex` can be either `index` (refcount 2) or the result of `PyLong_FromSsize_t` (refcount 1). In either case, the caller (`SetItem`) will decref, so everything is fine. I will add a comment about this!